### PR TITLE
Improve mnist-mlp to support for CUDA

### PR DESCRIPTION
**Pull request by junjihashimoto**
This pr fixes https://github.com/hasktorch/hasktorch/issues/501.
Fixed mnist-mlp to support CUDA.

### CODE REVIEW

File: examples/mnist-mlp/Main.hs [hunk 1]
```diff
@@ -1,86 +1,97 @@
...
-          (newParam, _) <- runStep model optimizer loss 1e-3
-          pure newParam
-        done = pure
-        begin = pure model
+trainLoop :: Optimizer o => Device -> MLP -> o -> ListT IO (Tensor, Tensor) -> IO MLP
+trainLoop localDevice model optimizer = P.foldM step begin done . enumerateData
+  where
+    step :: MLP -> ((Tensor, Tensor), Int) -> IO MLP
+    step model args = do
+      let ((input, label), iter) = toDevice localDevice args
```
Comment (junjihashimoto) [inline examples/mnist-mlp/Main.hs:62]: This line sends the training data to CUDA.

**Comment by austinvhuang** [comment]
Thanks @junjihashimoto ! Any idea what's going on with the stack overflow exceptions in CI?

**Comment by junjihashimoto** [comment]
I think the problem is the implementation of _asTensor. It's being fixed now.

**Review by tscholak — APPROVED** [review]
LGTM, thanks @junjihashimoto!

File: hasktorch/src/Torch/Tensor.hs [hunk 2]
```diff
@@ -453,10 +453,16 @@ instance (TensorIndex a, TensorIndex b, TensorIndex c, TensorIndex d, TensorInde
 -- Scalar <-> Tensor promotion
 --------------------------------------------------------------------------------
 
+asValue :: TensorLike a => Tensor -> a
+asValue t =
+  let t' = if device t == Device CPU 0 then asValue t else asValue (toCPU t)
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Tensor.hs:458]: asValue of this line is not needed.

**Comment by junjihashimoto** [comment]
Thank you, too!

**Pull request merged**